### PR TITLE
Introduce Cloneable interface

### PR DESCRIPTION
**Pull request by privat**
As exposed by #917 a standard API to clone objects could be useful.

This PR propose a Cloneable interface that class can implement.
The specif semantic of what a clone is is domain-dependent, the only suggestion I define is to be POLA and to consider the post condition `self == clone`. It seems sane since the precise semantic of `==` is also domain-dependent. The post-condition is not mandatory since not all classes redefines `==` and maybe some classes could be cloneable without being `==`-able. The reason is that `clone` is instantaneous but the semantic `==` should remain valid all the time. For instance, It could make sense to clone some Iterators but I am not sure what is a sane definition of `==` on them.

As a POC, clone is implemented is the three Array-related classes, Array, ArraySet and ArrayMap.

### CODE REVIEW

**Comment by R4PaSs** [comment]
C'est peut-être moi, mais me semble `Cloneable` serait plus juste grammaticalement ?
Ça serait le même nom qu'en [Java](http://docs.oracle.com/javase/8/docs/api/java/lang/Cloneable.html) et [C#](https://msdn.microsoft.com/en-us/library/system.icloneable%28v=vs.110%29.aspx) au passage, ça perdrait moins les gens qui arrivent dans le langage comme ça.

**Comment by R4PaSs** [comment]
+1 ça me va !

File: lib/standard/collection/abstract_collection.nit [hunk 1]
```diff
@@ -466,6 +466,30 @@ interface MapRead[K, V]
 	# Note: the value is returned *as is*, implementations may want to store the value in the map before returning it
 	# @toimplement
 	protected fun provide_default_value(key: K): V do abort
+
+	# Does `self` and `other` has the same keys associated with the same values?
```
Comment (Morriar) [inline lib/standard/collection/abstract_collection.nit]: have*

File: lib/standard/collection/array.nit [hunk 2]
```diff
@@ -252,6 +252,7 @@ end
 #     assert a == b
 class Array[E]
 	super AbstractArray[E]
+	super Clonable
```
Comment (Morriar) [inline lib/standard/collection/array.nit]: Cloneable?

File: lib/standard/kernel.nit [hunk 3]
```diff
@@ -226,6 +226,17 @@ interface Discrete
 	end
 end
 
+# Something that can be cloned
+#
+# The main method on this interface is `clone` and the specific
```
Comment (Morriar) [inline lib/standard/kernel.nit]: Suggestion:

~~~nit
# Something that can be cloned.
# 
# This interface introduces the `clone` method used to duplicate an instance.
interface Clonable
    # Duplicate `self`.
    # 
    # The specific semantic of this method is let to the subclasses.
    # For example you can choose that the cloning is done in depth by cloning
    # the linked instances or only shallow.
    #
    # See `Array::clone` for a shallow implementation example.
    # See `??` for an in depth clone example.
    #
    # Note: as the returned clone depends on the semantic, one should also want to redefine the
    # `==` method to specify the equality between an object and his clone.
    fun clone: SELF is abstract
end
~~~

**Comment by Morriar** [comment]
+1

**Comment by privat** [comment]
reroll: fix spelling and improve doc (inspired by the proposition of @Morriar but with a descriptive style)

**Comment by jpages** [comment]
+1

File: lib/standard/kernel.nit [hunk 4]
```diff
@@ -226,6 +226,25 @@ interface Discrete
 	end
 end
 
+# Something that can be cloned
+#
+# This interface introduces the `clone` method used to duplicate an instance
+# Its the specific semantic is let to the subclasses.
```
Comment (Morriar) [inline lib/standard/kernel.nit]: Its the?

**Comment by Morriar** [comment]
Stills a typo but the comment is better. Good job

**Comment by xymus** [comment]
+1

**Pull request merged**